### PR TITLE
Replace RefineTypes with dtype functions

### PR DESCRIPTION
**Pull request by ramiro050**
This commit adds dtype functions for all the torch ops that did not previously have one and removes the pass `RefineTypes`, since the abstract interpretation library now takes care of all the dtype propagation.

All dtype functions added are tested except for
- `aten.embedding`
- `aten._embedding_bag`
- `aten.embedding_bag`

These functions need a change to the testing framework to allow specifying the actual data inside the tensor used for testing. I will fix this in a follow up patch.

Co-authored-by: Jiahao Li <liplus17@163.com>

### CODE REVIEW

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py [hunk 1]
```diff
@@ -79,27 +83,6 @@ def aten〇exp〡shape(self: List[int]) -> List[int]:
...
-    Invocation(NonZeroDTensorWithDtype(torch.int32)),
-    Invocation(NonZeroDTensorWithDtype(torch.bool)),
-    Invocation(ZeroDTensorWithDtype(torch.float32)),
-    Invocation(ZeroDTensorWithDtype(torch.float64)),
-    Invocation(ZeroDTensorWithDtype(torch.bfloat16)),
-    Invocation(ZeroDTensorWithDtype(torch.int64)),
-    Invocation(ZeroDTensorWithDtype(torch.int32)),
-    Invocation(ZeroDTensorWithDtype(torch.bool)),
-])
-def aten〇expm1〡dtype(self_rank_dtype: Tuple[int, int]) -> int:
```
Comment (ramiro050) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py:96]: Removed dtype functions are just shifted down to be grouped with all the other dtype functions

File: lib/Dialect/Torch/Transforms/ReifyAbstractInterpCalculationsUtils.cpp [hunk 2]
```diff
@@ -176,20 +176,22 @@ FailureOr<Value> Torch::adjustFunctionArg(
...
+              .isa<Torch::IntType, Torch::FloatType, Torch::NoneType>();
         }))
       return b.create<DerefineOp>(loc, desiredType, operand).getResult();
   }
 
   // If the operand is NoneType, then we just need to derefine it to the
   // optional type in the function signature.
   if (operandType.isa<Torch::NoneType>()) {
-    assert(desiredType.isa<Torch::OptionalType>() &&
+    assert(!desiredType.isa<Torch::NoneType>() &&
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/ReifyAbstractInterpCalculationsUtils.cpp:194]: Can you update the comment here?

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/library_generator.py [hunk 3]
```diff
@@ -14,6 +14,55 @@
...
+    return [torch.float16, torch.bfloat16, torch.float32, torch.float64]
+
+def is_float_dtype(dtype: int) -> bool:
+    return dtype in all_float_dtypes()
+
+def get_priority_of_dtype(dtype: int) -> int:
+    # If a loop is used to iterate over a list of sorted dtypes, TorchScript
+    # produces a loop with INT64_MAX max trip count, which causes problems
+    # during the loop unrolling that takes place when simplifying the dtype
+    # functions. Therefore, here we result to `if`s.
```
Comment (silvasean) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/library_generator.py:39]: nit: s/result/resort/

**Review by silvasean — APPROVED** [review]
Nice work landing this! Just a few nits :rocket:

**Comment by silvasean** [comment]
Big thanks to Jiahao Li and any others that helped with this migration! :heart:

**Pull request merged**